lua/nvim-treesitter/textobjects/select.lua

```diff
@@ -134,11 +134,18 @@ function M.detect_selection_mode(query_string, keymap_mode)
     selection_mode = selection_modes or "v"
   end
 
-  if selection_mode == "n" then
-    selection_mode = "v"
+  -- According to "mode()" mapping, if we are in operator pending mode or visual mode,
+  -- then last char is {v,V,<C-v>}, exept for "no", which is "o", in which case we honor
+  -- last set `selection_mode`
+  local mode = vim.fn.mode(1)
+  local is_normal_or_charwise_v = mode == "n" or mode == "v"
+
+  if is_normal_or_charwise_v then
+    return mode == "n" and "v" or mode == "v" and selection_mode
```

Comment:
I would just write it case by case like:

```lua
if mode == "n" then
  return "v"
end

if mode == "v" then
  return selection_mode
end
```

However, what's the reason we'd use `v` mode in normal mode? I think either way it should be `selection_mode`

Reply:
I think we should first set the selection_mode to visual_mode?

Reply:
> However, what's the reason we'd use `v` mode in normal mode? I think either way it should be `selection_mode`

It's you who added that haha (#420).

Reply:
Also, assigning a parameter to `vim.fn.mode()` function seems to be no longer working in `nvim nightly`. It just returns on character regardless of the existence of parameter

Reply:
What about we do something like this? 
```lua
  -- ...
  local selection_mode = vim.fn.mode()
  if type(selection_modes) == "table" then
    selection_mode = selection_modes[query_string] or selection_mode
  elseif selection_modes then
    selection_mode = selection_modes
  end

  if selection_mode == "n" then
    selection_mode = "v"
  end

  return selection_mode
```

Reply:
> > However, what's the reason we'd use `v` mode in normal mode? I think either way it should be `selection_mode`
> 
> It's you who added that haha (#420).

That logic is different because it is stopping from accidentally returning "n" as a selection mode. In your logic, you're looking at the current vim mode, but not `selection_mode` configured in this plugin.

Reply:
Oh, my bad. Here is almost 1AM. Will check it back tomorrow morning when I have rested 😬.

Reply:
Although it seems to work, because if `selection_mode` is `n`, so will be `mode`, no? I tried `:TSTextobjectSelect @function.outer` and it selected correctly. In what other case would it fail?

Reply:
> Oh, my bad. Here is almost 1AM. Will check it back tomorrow morning when I have rested 😬.

Take good care of yourself! :)

Reply:
> Also, assigning a parameter to vim.fn.mode() function seems to be no longer working in nvim nightly. It just returns on character regardless of the existence of parameter

I just downloaded the lastest nightly and it works. Try something like `y:echo mode(1)`.